aws/aws_secrets_manager/aws_scm.go

```diff
@@ -0,0 +1,220 @@
+		region = os.Getenv(utils.AwsRegionKey)
+		if region == "" {
+			return nil, utils.ErrAWSRegionNotProvided
+		}
+	}
+
+	id, secret, token, err := utils.AuthKeys(secretConfig)
+	if err != nil {
+		return nil, err
+	}
+	asc, err := sc.NewAWSCredentials(id, secret, token, true)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get credentials: %v", err)
+	}
+	creds, err := asc.Get()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get credentials: %v", err)
+	}
+	config := &aws.Config{
+		Credentials: creds,
+		Region:      &region,
+	}
+	sess, err := session.NewSession(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a session: %v", err)
+	}
+	scm := secretsmanager.New(sess)
+	return &awsSecretsMgr{
+		scm:   scm,
+		sess:  sess,
+		creds: creds,
+		asc:   asc,
+	}, nil
+}
+
+func (a *awsSecretsMgr) String() string {
+	return Name
+}
+
+func (a *awsSecretsMgr) GetSecret(
+	secretId string,
+	keyContext map[string]string,
+) (map[string]interface{}, error) {
+	secretValueOutput, err := a.scm.GetSecretValue(&secretsmanager.GetSecretValueInput{
+		SecretId: aws.String(secretId),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == secretsmanager.ErrCodeResourceNotFoundException {
+				return nil, secrets.ErrInvalidSecretId
+			} else if aerr.Code() == secretsmanager.ErrCodeInvalidRequestException &&
+				strings.Contains(aerr.Error(), "marked for deletion") {
+				return nil, secrets.ErrInvalidSecretId
+			}
+		}
+		return nil, &secrets.ErrProviderInternal{Reason: err.Error(), Provider: Name}
+	}
+	if secretValueOutput.SecretString == nil || len(*secretValueOutput.SecretString) <= 0 {
+		return nil, secrets.ErrEmptySecretData
+	}
+	secretOut := make(map[string]interface{})
+	if err := json.Unmarshal([]byte(*secretValueOutput.SecretString), &secretOut); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal secret data: %v", err)
+	}
+	return secretOut, nil
+}
+
+func (a *awsSecretsMgr) PutSecret(
+	secretId string,
+	secretData map[string]interface{},
+	keyContext map[string]string,
+) error {
+	// Marshal the secret data
+	secretBytes, err := json.Marshal(secretData)
+	if err != nil {
+		return fmt.Errorf("failed to marshal secret data: %v", err)
+	}
+	// Check if there already exists a key.
+	_, err = a.scm.GetSecretValue(&secretsmanager.GetSecretValueInput{
+		SecretId: aws.String(secretId),
+	})
+	if err == nil {
+		// Update the existing secret
+		_, err = a.scm.PutSecretValue(&secretsmanager.PutSecretValueInput{
+			SecretId:     aws.String(secretId),
+			SecretString: aws.String(string(secretBytes)),
+		})
+	} else {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == secretsmanager.ErrCodeResourceNotFoundException {
+				// Create a new secret
+				_, err = a.scm.CreateSecret(&secretsmanager.CreateSecretInput{
+					SecretString: aws.String(string(secretBytes)),
+					Name:         aws.String(secretId),
+				})
+			} // return the aws error
```

Comment:
are we returning error here if it's not ResourceNotFoundException?

Reply:
oh nvm, i see the logic here. The error are propagated down to line 145.

Reply:
should we remove the comment then?

Reply:
The comment right after an if closing statement indicates that the else condition will return back the error by this function